zookeeper-server/src/main/java/org/apache/zookeeper/common/X509Util.java

```diff
@@ -446,4 +458,119 @@ private void configureSSLServerSocket(SSLServerSocket sslServerSocket) {
+     */
+    public void enableCertFileReloading() throws IOException {
+        LOG.info("enabling cert file reloading");
+        ZKConfig config = new ZKConfig();
+        String keyStoreLocation = config.getProperty(sslKeystoreLocationProperty);
+        if (keyStoreLocation != null && !keyStoreLocation.isEmpty()) {
+            final Path filePath = Paths.get(keyStoreLocation).toAbsolutePath();
+            Path parentPath = filePath.getParent();
+            if (parentPath == null) {
+                throw new IOException(
+                        "Key store path does not have a parent: " + filePath);
+            }
+            FileChangeWatcher newKeyStoreFileWatcher = new FileChangeWatcher(
+                    parentPath,
+                    watchEvent -> {
+                        handleWatchEvent(filePath, watchEvent);
+                    });
+            // stop old watcher if there is one
+            if (keyStoreFileWatcher != null) {
+                keyStoreFileWatcher.stop();
+            }
+            keyStoreFileWatcher = newKeyStoreFileWatcher;
+            keyStoreFileWatcher.start();
+        }
+        String trustStoreLocation = config.getProperty(sslTruststoreLocationProperty);
+        if (trustStoreLocation != null && !trustStoreLocation.isEmpty()) {
+            final Path filePath = Paths.get(trustStoreLocation).toAbsolutePath();
+            Path parentPath = filePath.getParent();
+            if (parentPath == null) {
+                throw new IOException(
+                        "Trust store path does not have a parent: " + filePath);
+            }
+            FileChangeWatcher newTrustStoreFileWatcher = new FileChangeWatcher(
+                    parentPath,
+                    watchEvent -> {
+                        handleWatchEvent(filePath, watchEvent);
+                    });
+            // stop old watcher if there is one
+            if (trustStoreFileWatcher != null) {
+                trustStoreFileWatcher.stop();
+            }
+            trustStoreFileWatcher = newTrustStoreFileWatcher;
+            trustStoreFileWatcher.start();
+        }
+    }
+
+    /**
+     * Disables automatic reloading of the trust store and key store files when they change on disk.
+     * Stops background threads and closes WatchService instances.
+     */
+    public void disableCertFileReloading() {
+        if (keyStoreFileWatcher != null) {
+            keyStoreFileWatcher.stop();
+            keyStoreFileWatcher = null;
+        }
+        if (trustStoreFileWatcher != null) {
+            trustStoreFileWatcher.stop();
+            trustStoreFileWatcher = null;
+        }
+    }
+
+    // Finalizer guardian object, see Effective Java item 7
+    // TODO: finalize() is deprecated starting with Java 10. This needs to be
+    // replaced with an explicit shutdown call.
+    @SuppressWarnings("unused")
+    private final Object finalizerGuardian = new Object() {
```

Comment:
Reading the referenced literature about this, I believe it should be better to avoid using finalizer like this. We might even be better avoid using finalizers entirely. There's no guarantee when finalizer gets executed, not even guarantee to be executed at all, it has a huge performance penalty, etc.

We should rather implement an explicit `close()` method with the `AutoClosable` interface and call it from QuorumPeer's shutdown() method.

Reply:
I'm worried about forgetting to call `close()` and leaking the background threads. X509Util is created in other places besides QuorumPeer. But I'll see what I can do about it, we only need to call close() if we called `enableCertFileReloading()` so it might be ok ...

Reply:
I got your point, but in my understanding this is not the recommended way to do destructor-like behaviour in Java. The article advice on using the guardian pattern only for a safety net besides the `close()` method, logging a warning message if the class has not been properly closed by the user. This could be a valid case for us too, but given that this is purely our codebase, not a public API, I think it should be preferrable to expect that `X509Util` is closed properly everywhere.

Reply:
Done.